### Multiple Linear Regression


In [1]:
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
import seaborn as sn


In [2]:
data=pd.read_csv("1.02.+Multiple+linear+regression.csv")
data

SAT   GPA  Rand 1,2,3
0   1714  2.40           1
1   1664  2.52           3
2   1760  2.54           3
3   1685  2.74           3
4   1693  2.83           2
..   ...   ...         ...
79  1936  3.71           3
80  1810  3.71           1
81  1987  3.73           3
82  1962  3.76           1
83  2050  3.81           2

[84 rows x 3 columns]

In [3]:
data.describe()

SAT        GPA  Rand 1,2,3
count    84.000000  84.000000   84.000000
mean   1845.273810   3.330238    2.059524
std     104.530661   0.271617    0.855192
min    1634.000000   2.400000    1.000000
25%    1772.000000   3.190000    1.000000
50%    1846.000000   3.380000    2.000000
75%    1934.000000   3.502500    3.000000
max    2050.000000   3.810000    3.000000

In [4]:
data.head()

SAT   GPA  Rand 1,2,3
0  1714  2.40           1
1  1664  2.52           3
2  1760  2.54           3
3  1685  2.74           3
4  1693  2.83           2

In [5]:
x=data[["SAT","Rand 1,2,3"]]
y=data["GPA"]

In [6]:
from sklearn.linear_model import LinearRegression

In [7]:
reg=LinearRegression()

In [8]:
reg.fit(x,y)

LinearRegression()

In [9]:
reg.coef_

array([ 0.00165354, -0.00826982])

In [10]:
# calculation of Rsquared
reg.score(x,y)

0.4066811952814282

### FORMULA FOR ADJUSTED R^2
$R^2_{adj.} = 1 - (1-R^2)*\frac{n-1}{n-p-1}$

 n=84(number of observations)
 
p=2(number of  predictors)

In [11]:
x.shape

(84, 2)

In [12]:
r2=reg.score(x,y)
n=x.shape[0]
p=x.shape[1]
adjusted_r2=1-(1-r2)*(n-1)/(n-p-1)
adjusted_r2

0.39203134825134

# FEATURE SELECTION

 feature selection models, improves speed and prevents a series of unwanted issues arisisng from having too many features
 
feature selection main hum model ki values ko identify karte hai ki wo model ke prediction ke liye kitni important hai

In [13]:
from sklearn.feature_selection import f_regression

In [14]:
f_regression(x,y)

(array([56.04804786,  0.17558437]), array([7.19951844e-11, 6.76291372e-01]))

#### f-statistice- values (array([56.04804786,  0.17558437]),

#### p-values array([7.19951844e-11, 6.76291372e-01]))

In [15]:
p_values=f_regression(x,y)[1]
p_values

array([7.19951844e-11, 6.76291372e-01])

In [16]:
p_values.round(3)


array([0.   , 0.676])

here SAT value is 0. and Rand 1,2,3 is 0.676 that means the value of Rand 1,2,

### creating a summary table


In [17]:
reg_summary=pd.DataFrame(data=x.columns.values,columns=['Features'])
reg_summary

Features
0         SAT
1  Rand 1,2,3

In [18]:
reg_summary ["coefficients"]=reg.coef_
reg_summary["p-values"]=p_values.round(3)

In [19]:
reg_summary

Features  coefficients  p-values
0         SAT      0.001654     0.000
1  Rand 1,2,3     -0.008270     0.676

# standardization

$ Standard  variable = \frac{x-u}{SIGMA}$

 - x - original varibale
 - u - mean of original variable
 - sigma - standard deviation of original variable

In [55]:
yy=data['SAT']
yy

0     1714
1     1664
2     1760
3     1685
4     1693
      ... 
79    1936
80    1810
81    1987
82    1962
83    2050
Name: SAT, Length: 84, dtype: int64

- import statistics as st
- st.mean(yy)
- st
-  st.stdev(yy)


In [20]:
from sklearn.preprocessing import StandardScaler


In [21]:
scalar=StandardScaler()

 scalar will be used to subtract the mean and divide by the standard deviation

In [22]:
scalar.fit(x)

StandardScaler()

 fit calculate and stores the mean and standard deviation of each feature

In [23]:
x_scaled=scalar.transform(x)

scalar.transform transform the unscaled inputs using the information contained in the scaler object(feature-wise)
it subtract the mean and divide standard deviation with each feature

In [24]:
x_scaled

array([[-1.26338288, -1.24637147],
       [-1.74458431,  1.10632974],
       [-0.82067757,  1.10632974],
       [-1.54247971,  1.10632974],
       [-1.46548748, -0.07002087],
       [-1.68684014, -1.24637147],
       [-0.78218146, -0.07002087],
       [-0.78218146, -1.24637147],
       [-0.51270866, -0.07002087],
       [ 0.04548499,  1.10632974],
       [-1.06127829,  1.10632974],
       [-0.67631715, -0.07002087],
       [-1.06127829, -1.24637147],
       [-1.28263094,  1.10632974],
       [-0.6955652 , -0.07002087],
       [ 0.25721362, -0.07002087],
       [-0.86879772,  1.10632974],
       [-1.64834403, -0.07002087],
       [-0.03150724,  1.10632974],
       [-0.57045283,  1.10632974],
       [-0.81105355,  1.10632974],
       [-1.18639066,  1.10632974],
       [-1.75420834,  1.10632974],
       [-1.52323165, -1.24637147],
       [ 1.23886453, -1.24637147],
       [-0.18549169, -1.24637147],
       [-0.5608288 , -1.24637147],
       [-0.23361183,  1.10632974],
       [ 1.68156984,

# Regression with feature scaling

In [25]:
reg=LinearRegression()
reg.fit(x_scaled,y)


LinearRegression()

In [26]:
reg.coef_

array([ 0.17181389, -0.00703007])

In [27]:

reg.intercept_

3.330238095238095

In [59]:
reg.score(x_scaled,y)

0.4066811952814283

### creating a summary table

In [28]:

reg_summary=pd.DataFrame([["Intercept"],["SAT"],["Rand 1,2,3"]],columns=["features"])
reg_summary["weights"]=reg.intercept_,reg.coef_[0],reg.coef_[1]

In [29]:
reg_summary

features   weights
0   Intercept  3.330238
1         SAT  0.171814
2  Rand 1,2,3 -0.007030

bigger the weight than bigger the impact of regression

# making predictions with standardized coefficients

In [30]:
new_data=pd.DataFrame(data=[[1700,2],[1800,1]],columns=["SAT","Rand 1,,2,3"])
new_data

SAT  Rand 1,,2,3
0  1700            2
1  1800            1

In [31]:
reg.predict(new_data)

array([295.39979563, 312.58821497])

In [32]:
new_data_scaled=scalar.transform(new_data)
new_data_scaled

array([[-1.39811928, -0.07002087],
       [-0.43571643, -1.24637147]])

In [33]:
reg.predict(new_data_scaled)

array([3.09051403, 3.26413803])

### what if we remove 'Random 1,2,3 variable?

In [34]:
reg_simple=LinearRegression()
x_simple_matrix=x_scaled[:,0].reshape(-1,1)
reg_simple.fit(x_simple_matrix,y)

LinearRegression()

In [35]:
reg_simple.predict(new_data_scaled[:,0].reshape(-1,1))

array([3.08970998, 3.25527879])

### overfitting and underfiting

underfiting - the model has not captured the underlying logic of the data 
 
 overfitting - our training has focused on the particular training set so much it has missed the pointmeans that the predict model jo hai vo actual values ke under fit hai yani sare point ko attach kar raha hai
 
 overfiting is close to the actual value and it is good predictive but underfitting is low predictive power and low accuracy

# Train Test Split

In [36]:
import numpy as np
from sklearn.model_selection import train_test_split


In [37]:
a=np.arange(1,101)
a

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100])

In [38]:
b=np.arange(501,601)
b

array([501, 502, 503, 504, 505, 506, 507, 508, 509, 510, 511, 512, 513,
       514, 515, 516, 517, 518, 519, 520, 521, 522, 523, 524, 525, 526,
       527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 538, 539,
       540, 541, 542, 543, 544, 545, 546, 547, 548, 549, 550, 551, 552,
       553, 554, 555, 556, 557, 558, 559, 560, 561, 562, 563, 564, 565,
       566, 567, 568, 569, 570, 571, 572, 573, 574, 575, 576, 577, 578,
       579, 580, 581, 582, 583, 584, 585, 586, 587, 588, 589, 590, 591,
       592, 593, 594, 595, 596, 597, 598, 599, 600])

### split the data

In [39]:
train_test_split(a)

[array([ 48,  96,  28,  95,  31, 100,  93,  61,   3,  38,  82,   8,   6,
          1,  42,  37,  64,  41,  19,  56,  45,  77,  16,  25,  54,  74,
          5,  55,  70,  36,  39,  81,   4,  57,  72,  92,  94,  66,  85,
         13,  32,  79,  46,  17,  68,   9,  87,  50,  67,  29,  86,  22,
         58,  99,  12,  26,  15,  21,  33,  44,  30,  97,  83,  52,  24,
         59,  40,  73,  90,  89,  62,   2,  69,  49,  20]),
 array([53, 47, 71, 18, 10, 80, 35, 88, 84, 14, 78, 27, 98,  7, 34, 65, 23,
        63, 60, 43, 76, 91, 75, 11, 51])]



{###TRAINING ARRAT}  [array([ 98,  24,  89,  12,  94,  78,  59,  35,   7,  96,  79,  56,  53,
         40,   8,  17,  55,   6,  65,  82,  48,  26,  68,   9,  70,  88,
         46,  51,  71,  84,  27,  42,  85,   2,  18,  22,  16,  90, 100,
         69,  99,  20,  72,  52,  83,  66,  39,  38,  75,  36,  33,  34,
         63,  54,  32,  76,  29,  45,  15,  57,  73,   3,  91,  10,  31,
         25,  93,   5,  13,  21,  49,  30,  77,  95,  41]),


{# TESTING ARRAY}array([50, 60, 23,  1, 74, 81, 43, 64, 80, 61, 86, 62, 11, 47, 37, 97, 58,
        92, 67,  4, 28, 19, 87, 44, 14])]

In [40]:
# a_train,a_test=train_test_split(a)
a_train,a_test=train_test_split(a,test_size=0.2)

In [41]:
a_train

array([99, 29, 36, 64, 79, 53, 90, 49, 77, 31, 46,  9, 82, 71, 20, 75, 38,
       48, 69, 68, 58, 54, 40, 67, 98, 17, 78, 91, 33, 97,  1, 96, 18, 72,
       30,  8, 37, 43, 83,  3, 22,  5,  4, 24, 74, 21, 66, 62, 89, 85, 63,
       23, 44, 35, 52, 87, 47, 13, 65, 95, 86,  2, 60, 81, 15, 57, 16, 34,
       41, 88, 26, 93, 19, 11, 14, 50,  6, 45,  7, 92])


a_train,a_test=train_test_split(a,test_size=0.2,shuffle=False)

a_test

array([ 81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,  92,  93,
        94,  95,  96,  97,  98,  99, 100])

### Explore the result

In [42]:
# # to stop the shuffling of the data 
# a_train,a_test=train_test_split(a,test_size=0.2,shuffle=False)
a_train,a_test=train_test_split(a,test_size=0.2)

In [43]:
a_train,a_test,b_train,b_test=train_test_split(a,b,test_size=0.2,random_state=42)
# it can be different shuffle by adding different random_state

In [44]:
a_train.shape,a_test.shape

((80,), (20,))

In [45]:
a_train

array([ 56,  89,  27,  43,  70,  16,  41,  97,  10,  73,  12,  48,  86,
        29,  94,   6,  67,  66,  36,  17,  50,  35,   8,  96,  28,  20,
        82,  26,  63,  14,  25,   4,  18,  39,   9,  79,   7,  65,  37,
        90,  57, 100,  55,  44,  51,  68,  47,  69,  62,  98,  80,  42,
        59,  49,  99,  58,  76,  33,  95,  60,  64,  85,  38,  30,   2,
        53,  22,   3,  24,  88,  92,  75,  87,  83,  21,  61,  72,  15,
        93,  52])

In [46]:
a_test

array([84, 54, 71, 46, 45, 40, 23, 81, 11,  1, 19, 31, 74, 34, 91,  5, 77,
       78, 13, 32])

In [47]:
b_train

array([556, 589, 527, 543, 570, 516, 541, 597, 510, 573, 512, 548, 586,
       529, 594, 506, 567, 566, 536, 517, 550, 535, 508, 596, 528, 520,
       582, 526, 563, 514, 525, 504, 518, 539, 509, 579, 507, 565, 537,
       590, 557, 600, 555, 544, 551, 568, 547, 569, 562, 598, 580, 542,
       559, 549, 599, 558, 576, 533, 595, 560, 564, 585, 538, 530, 502,
       553, 522, 503, 524, 588, 592, 575, 587, 583, 521, 561, 572, 515,
       593, 552])

In [48]:
b_test

array([584, 554, 571, 546, 545, 540, 523, 581, 511, 501, 519, 531, 574,
       534, 591, 505, 577, 578, 513, 532])